File: deadband-fortran/test_deadband.f90
program test_deadband
  use deadband
  implicit none

  integer :: pass_count, fail_count, total_tests
  real(8) :: t_start, t_end
  logical :: ok

  pass_count = 0
  fail_count = 0
  total_tests = 0

  print '(a)', '========================================================='
  print '(a)', '  DEADBAND FRAMEWORK -- Fortran Test Suite'
  print '(a)', '========================================================='
  print '(a)', ''

  ! ---- Constants ----
  call test_approx('PHI = 1.6180339887...', PHI, 1.6180339887498948d0)
  call test_approx('PHI_INV = 0.6180339887...', PHI_INV, 0.6180339887498948d0)
  call test_approx('SQRT3 = 1.7320508075...', SQRT3, 1.7320508075688772d0)

  ! ---- Fibonacci ----
  call test_int('Fibonacci(0) = 0', int(fibonacci(0)), 0)
  call test_int('Fibonacci(1) = 1', int(fibonacci(1)), 1)
  call test_int('Fibonacci(10) = 55', int(fibonacci(10)), 55)
  call test_int('Fibonacci(20) = 6765', int(fibonacci(20)), 6765)

  ! ---- Eisenstein Snap ----
  print '(a)', '--- Eisenstein Snap ---'

  call cpu_time(t_start)
  block
    type(SnapResult) :: sr
    sr = eisenstein_snap(1.0d0, 0.0d0)
    call test_approx('Snap (1,0) → (1,0)', sr%x, 1.0d0)
    call test_approx('Snap (1,0) error=0', sr%error, 0.0d0)

    sr = eisenstein_snap(0.5d0, SQRT3/2.0d0)
    call test_approx('Snap (0.5, √3/2) → (-0.5, √3/2)', sr%x, -0.5d0)
    call test_approx('Snap (0.5, √3/2) y', sr%y, SQRT3/2.0d0)

    sr = eisenstein_snap(0.3d0, 0.3d0)
    call test_true('Snap (0.3,0.3) error small', sr%error < 0.5d0)

    sr = eisenstein_snap(0.0d0, 0.0d0)
    call test_approx('Snap origin → origin', sr%error, 0.0d0)
  end block
  call cpu_time(t_end)
  print '(a,f8.4,a)', '  Eisenstein snap time: ', t_end - t_start, 's'

  ! ---- HPDF Sampling ----
  print '(a)', ''
  print '(a)', '--- HPDF Sampling ---'
  call cpu_time(t_start)
  block
    real(8) :: sx, sy
    integer :: i
    do i = 1, 10000
      call hpdf_sample(sx, sy)
    end do
    call test_true('HPDF 10k samples completed', .true.)
  end block
  call cpu_time(t_end)
  print '(a,f8.4,a)', '  HPDF 10k samples time: ', t_end - t_start, 's'

  ! ---- /360 Arithmetic ----
  print '(a)', ''
  print '(a)', '--- /360 Field Arithmetic ---'
  call test_int8('/360 100+200=300', div360_add(100_8, 200_8), 300_8)
  call test_int8('/360 200+200=40', div360_add(200_8, 200_8), 40_8)
  call test_int8('/360 50-100=310', div360_sub(50_8, 100_8), 310_8)
  call test_int8('/360 180*2=0', div360_mul(180_8, 2_8), 0_8)
  call test_int8('/360 90*3=270', div360_mul(90_8, 3_8), 270_8)
  call test_int8('/360 359+1=0', div360_add(359_8, 1_8), 0_8)
  call test_int8('/360 0-1=359', div360_sub(0_8, 1_8), 359_8)

  ! ---- BMA over GF(2) ----
  print '(a)', ''
  print '(a)', '--- BMA (Berlekamp-Massey over GF(2)) ---'
  call cpu_time(t_start)
  block
    integer :: seq1(7)
    integer :: ln

    ! Simple: 1,0,1,0,1,0,1 → LFSR length 2 (toggles)
    seq1 = [1, 0, 1, 0, 1, 0, 1]
    ln = bma_detect(seq1, 7)
    call test_int('BMA [1,0,1,0,1,0,1] len=2', ln, 2)

    ! All zeros → length 0
    seq1 = [0, 0, 0, 0, 0, 0, 0]
    ln = bma_detect(seq1, 7)
    call test_int('BMA [0,0,0,...] len=0', ln, 0)

    ! All ones → length 1
    seq1 = [1, 1, 1, 1, 1, 1, 1]
    ln = bma_detect(seq1, 7)
    call test_int('BMA [1,1,1,...] len=1', ln, 1)
  end block
  call cpu_time(t_end)
  print '(a,f8.4,a)', '  BMA time: ', t_end - t_start, 's'

  ! ---- Shell Eigenstructure ----
  print '(a)', ''
  print '(a)', '--- Shell Eigenstructure ---'
  call cpu_time(t_start)
  block
    real(8) :: cov(2,2), ke, ae, ratio
    integer :: st

    ! Identity matrix
    cov = reshape([1.0d0, 0.0d0, 0.0d0, 1.0d0], [2,2])
    call shell_decompose(cov, ke, ae, ratio, st)
    call test_int('Identity status=0', st, 0)
    call test_approx('Identity known_energy=2', ke, 2.0d0)
    call test_approx('Identity ratio=0.5', ratio, 0.5d0)

    ! Diagonal: eigenvalues 4 and 1
    cov = reshape([4.0d0, 0.0d0, 0.0d0, 1.0d0], [2,2])
    call shell_decompose(cov, ke, ae, ratio, st)
    call test_approx('Diag(4,1) known_energy=5', ke, 5.0d0)
    call test_approx('Diag(4,1) ratio=0.8', ratio, 0.8d0)

    ! Mixed with off-diagonal
    cov = reshape([2.0d0, 1.0d0, 1.0d0, 2.0d0], [2,2])
    call shell_decompose(cov, ke, ae, ratio, st)
    call test_approx('Mixed known_energy=4', ke, 4.0d0)
    call test_approx('Mixed ratio=0.75', ratio, 0.75d0)
  end block
  call cpu_time(t_end)
  print '(a,f8.4,a)', '  Eigenstructure time: ', t_end - t_start, 's'

  ! ---- Fibonacci-Spline Search ----
  print '(a)', ''
  print '(a)', '--- Fibonacci-Spline Search ---'
  call cpu_time(t_start)
  block
    integer, parameter :: N = 1000
    integer, parameter :: D = 3
    integer, parameter :: K = 5
    real(8) :: db(N, D), query(D)
    integer :: idx(K)
    real(8) :: sims(K)
    integer :: i

    ! Fill database with points
    do i = 1, N
      db(i, :) = real(i, 8) * [0.1d0, 0.2d0, 0.3d0]
    end do

    query = [50.0d0, 100.0d0, 150.0d0]  ! close to point 500

    call fib_spline_search(query, db, N, D, K, idx, sims)

    call test_true('Fib-spline found K results', count(idx > 0) >= 1)
    call test_true('Fib-spline similarities positive', all(sims > 0.0d0 .and. sims <= 1.0d0))
  end block
  call cpu_time(t_end)
  print '(a,f8.4,a)', '  Fib-spline search time: ', t_end - t_start, 's'

  ! ---- Dodecet Pack/Unpack ----
  print '(a)', ''
  print '(a)', '--- Dodecet Pack/Unpack (60-bit CDC Word) ---'
  block
    integer(8) :: word
    integer :: d0, d1, d2, d3, d4

    d0 = 100; d1 = 500; d2 = 1000; d3 = 2000; d4 = 4095
    word = pack_dodecets(d0, d1, d2, d3, d4)

    call test_int('Unpack dodecet 0', unpack_dodecet(word, 0), d0)
    call test_int('Unpack dodecet 1', unpack_dodecet(word, 1), d1)
    call test_int('Unpack dodecet 2', unpack_dodecet(word, 2), d2)
    call test_int('Unpack dodecet 3', unpack_dodecet(word, 3), d3)
    call test_int('Unpack dodecet 4', unpack_dodecet(word, 4), d4)
  end block

  ! ---- Summary ----
  print '(a)', ''
  print '(a)', '========================================================='
  write(*, '(a,i0,a,i0,a)') '  RESULTS: ', pass_count, ' PASS, ', fail_count, ' FAIL'
  print '(a)', '========================================================='

  if (fail_count > 0) stop 1

contains

  subroutine test_approx(name, actual, expected)
    character(*), intent(in) :: name
    real(8), intent(in) :: actual, expected
    real(8), parameter :: tol = 1.0d-10

    total_tests = total_tests + 1
    if (abs(actual - expected) < tol) then
      print '(a,a,a)', '  ✓ PASS: ', name, ''
      pass_count = pass_count + 1
    else
      print '(a,a,a,es12.4,a,es12.4)', '  ✗ FAIL: ', name, ' got ', actual, ' expected ', expected
      fail_count = fail_count + 1
    end if
  end subroutine

  subroutine test_int(name, actual, expected)
    character(*), intent(in) :: name
    integer, intent(in) :: actual, expected

    total_tests = total_tests + 1
    if (actual == expected) then
      print '(a,a)', '  ✓ PASS: ', name
      pass_count = pass_count + 1
    else
      print '(a,a,a,i0,a,i0)', '  ✗ FAIL: ', name, ' got ', actual, ' expected ', expected
      fail_count = fail_count + 1
    end if
  end subroutine

  subroutine test_int8(name, actual, expected)
    character(*), intent(in) :: name
    integer(8), intent(in) :: actual, expected

    total_tests = total_tests + 1
    if (actual == expected) then
      print '(a,a)', '  ✓ PASS: ', name
      pass_count = pass_count + 1
    else
      print '(a,a,a,i0,a,i0)', '  ✗ FAIL: ', name, ' got ', actual, ' expected ', expected
      fail_count = fail_count + 1
    end if
  end subroutine

  subroutine test_true(name, condition)
    character(*), intent(in) :: name
    logical, intent(in) :: condition

    total_tests = total_tests + 1
    if (condition) then
      print '(a,a)', '  ✓ PASS: ', name
      pass_count = pass_count + 1
    else
      print '(a,a)', '  ✗ FAIL: ', name
      fail_count = fail_count + 1
    end if
  end subroutine

end program test_deadband

File: deadband-c/test_deadband.c
#define _GNU_SOURCE
#include "deadband.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#define PASS(name) printf("  ✅ PASS  %s\n", name)
#define FAIL(name, ...) do { printf("  ❌ FAIL  %s — ", name); printf(__VA_ARGS__); printf("\n"); fails++; } while(0)

static double now_sec(void)
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec * 1e-9;
}

/* ─── 1. Eisenstein snap ──────────────────────────────────────── */

static int test_eisenstein(void)
{
    int fails = 0;

    /* Lattice point (a=1,b=0) → (1,0). Snap should give zero error. */
    SnapResult r = eisenstein_snap(1.0, 0.0);
    if (r.err > 1e-12 || r.a != 1 || r.b != 0)
        FAIL("snap(1,0)", "err=%.2e a=%ld b=%ld", r.err, (long)r.a, (long)r.b);
    else PASS("snap(1,0) = zero error");

    /* Lattice point (a=0,b=1) → (-0.5, √3/2) */
    r = eisenstein_snap(-0.5, 0.8660254037844386);
    if (r.err > 1e-10 || r.a != 0 || r.b != 1)
        FAIL("snap(-0.5,√3/2)", "err=%.2e a=%ld b=%ld", r.err, (long)r.a, (long)r.b);
    else PASS("snap(-0.5,√3/2) = zero error");

    /* Off-lattice point (0.3, 0.4) */
    r = eisenstein_snap(0.3, 0.4);
    if (r.err > 0.6)  /* rough upper bound */
        FAIL("snap(0.3,0.4) err too large", "err=%.4f", r.err);
    else PASS("snap(0.3,0.4) reasonable error");

    /* Origin */
    r = eisenstein_snap(0.0, 0.0);
    if (r.err > 1e-12 || r.a != 0 || r.b != 0)
        FAIL("snap(0,0)", "err=%.2e", r.err);
    else PASS("snap(0,0) = zero error");

    /* Many random snaps — verify error < max possible */
    srand(42);
    double max_err = 0.0;
    for (int i = 0; i < 10000; i++) {
        double x = (double)rand()/RAND_MAX * 10.0 - 5.0;
        double y = (double)rand()/RAND_MAX * 10.0 - 5.0;
        r = eisenstein_snap(x, y);
        if (r.err > max_err) max_err = r.err;
    }
    /* Max error should be < circumradius of hexagon ≈ 1/√3 ≈ 0.577 */
    if (max_err > 0.5774)
        FAIL("random snap max_err", "%.4f > 0.5774", max_err);
    else PASS("10000 random snaps max_err OK");

    return fails;
}

/* ─── 2. HPDF variance ────────────────────────────────────────── */

static int test_hpdf(void)
{
    int fails = 0;
    srand(12345);
    const int N = 100000;
    double sx = 0, sy = 0, sx2 = 0, sy2 = 0;

    double t0 = now_sec();
    for (int i = 0; i < N; i++) {
        Vec2 v = hpdf_sample();
        sx  += v.x;  sx2 += v.x * v.x;
        sy  += v.y;  sy2 += v.y * v.y;
    }
    double dt = now_sec() - t0;

    double mean_x = sx / N, mean_y = sy / N;
    double var_x  = sx2 / N - mean_x * mean_x;
    double var_y  = sy2 / N - mean_y * mean_y;

    /* Expected variance = 5/24 ≈ 0.208333 for regular hexagon with R=1 */
    double expected = 5.0 / 24.0;

    printf("  HPDF stats (%d samples, %.1f ms): mean=(%.4f,%.4f) var=(%.5f,%.5f) expected=%.5f\n",
           N, dt*1000, mean_x, mean_y, var_x, var_y, expected);

    if (fabs(mean_x) > 0.01)
        FAIL("HPDF mean_x not ~0", "%.4f", mean_x);
    else PASS("HPDF mean_x ≈ 0");

    if (fabs(mean_y) > 0.01)
        FAIL("HPDF mean_y not ~0", "%.4f", mean_y);
    else PASS("HPDF mean_y ≈ 0");

    if (fabs(var_x - expected) > 0.005)
        FAIL("HPDF var_x", "%.5f vs expected %.5f", var_x, expected);
    else PASS("HPDF var_x ≈ 5/24");

    if (fabs(var_y - expected) > 0.005)
        FAIL("HPDF var_y", "%.5f vs expected %.5f", var_y, expected);
    else PASS("HPDF var_y ≈ 5/24");

    return fails;
}

/* ─── 3. /360 arithmetic ──────────────────────────────────────── */

static int test_div360(void)
{
    int fails = 0;
    const int OPS = 100000;

    /* Zero-drift accumulator test */
    int64_t acc = 0;
    srand(999);
    double t0 = now_sec();
    for (int i = 0; i < OPS; i++) {
        int64_t v = rand() % 720 - 360;
        acc = div360_add(acc, v);
    }
    double dt = now_sec() - t0;
    printf("  /360: %d additions in %.1f ms, final acc = %ld\n",
           OPS, dt*1000, (long)acc);

    if (acc < 0 || acc >= 360)
        FAIL("div360 accumulator out of range", "%ld", (long)acc);
    else PASS("div360 accumulator in [0,359]");

    /* Specific edge cases */
    if (div360_add(350, 20) != 10)
        FAIL("div360_add(350,20)", "got %ld", (long)div360_add(350,20));
    else PASS("div360_add(350,20) = 10");

    if (div360_sub(10, 20) != 350)
        FAIL("div360_sub(10,20)", "got %ld", (long)div360_sub(10,20));
    else PASS("div360_sub(10,20) = 350");

    if (div360_mul(180, 2) != 0)
        FAIL("div360_mul(180,2)", "got %ld", (long)div360_mul(180,2));
    else PASS("div360_mul(180,2) = 0");

    if (div360_mul(-1, 180) != 180)
        FAIL("div360_mul(-1,180)", "got %ld", (long)div360_mul(-1,180));
    else PASS("div360_mul(-1,180) = 180");

    /* Batch drift test: repeated add-sub of same value must return 0 */
    int64_t val = 0;
    for (int i = 0; i < OPS; i++) {
        val = div360_add(val, 7);
        val = div360_sub(val, 7);
    }
    if (val != 0)
        FAIL("div360 drift test", "final=%ld", (long)val);
    else PASS("div360 100k add/sub pairs: zero drift");

    return fails;
}

/* ─── 4. Berlekamp-Massey ─────────────────────────────────────── */

static int test_bma(void)
{
    int fails = 0;

    /* All zeros → L = 0 */
    {
        uint8_t seq[20] = {0};
        int L = bma_detect(seq, 20);
        if (L != 0) FAIL("BMA all-zeros", "L=%d", L);
        else PASS("BMA all-zeros L=0");
    }

    /* Alternating 101010... → period 2, L=2 */
    {
        uint8_t seq[20];
        for (int i = 0; i < 20; i++) seq[i] = i & 1;
        int L = bma_detect(seq, 20);
        if (L != 2) FAIL("BMA alternating", "L=%d", L);
        else PASS("BMA alternating L=2");
    }

    /* All ones → L = 1 */
    {
        uint8_t seq[20];
        for (int i = 0; i < 20; i++) seq[i] = 1;
        int L = bma_detect(seq, 20);
        if (L != 1) FAIL("BMA all-ones", "L=%d", L);
        else PASS("BMA all-ones L=1");
    }

    /* Single 1 then zeros → L = 1 */
    {
        uint8_t seq[20] = {0};
        seq[0] = 1;
        int L = bma_detect(seq, 20);
        if (L > 1) FAIL("BMA 10000...", "L=%d", L);
        else PASS("BMA 10000... L<=1");
    }

    return fails;
}

/* ─── 5. Shell decompose ──────────────────────────────────────── */

static int test_shell(void)
{
    int fails = 0;

    /* Identity matrix → eigenvalues both 1.0, ratio 0.5 */
    {
        double cov[4] = {1, 0, 0, 1};
        ShellResult r = shell_decompose(cov);
        if (fabs(r.lam1 - 1.0) > 1e-10 || fabs(r.lam2 - 1.0) > 1e-10)
            FAIL("shell identity eigenvalues", "(%.4f, %.4f)", r.lam1, r.lam2);
        else PASS("shell identity eigenvalues = (1,1)");

        if (fabs(r.energy_ratio - 0.5) > 1e-10)
            FAIL("shell identity ratio", "%.4f", r.energy_ratio);
        else PASS("shell identity energy_ratio = 0.5");
    }

    /* Diagonal (3, 1) → eigenvalues 3 and 1, ratio 0.75 */
    {
        double cov[4] = {3, 0, 0, 1};
        ShellResult r = shell_decompose(cov);
        if (fabs(r.lam1 - 3.0) > 1e-10 || fabs(r.lam2 - 1.0) > 1e-10)
            FAIL("shell diag(3,1)", "(%.4f, %.4f)", r.lam1, r.lam2);
        else PASS("shell diag(3,1) eigenvalues correct");

        if (fabs(r.energy_ratio - 0.75) > 1e-10)
            FAIL("shell diag(3,1) ratio", "%.4f", r.energy_ratio);
        else PASS("shell diag(3,1) ratio = 0.75");
    }

    /* Non-diagonal with off-diagonal terms */
    {
        double cov[4] = {2, 1, 1, 2};
        ShellResult r = shell_decompose(cov);
        /* eigenvalues: 3 and 1 */
        if (fabs(r.lam1 - 3.0) > 1e-10 || fabs(r.lam2 - 1.0) > 1e-10)
            FAIL("shell [[2,1],[1,2]]", "(%.4f, %.4f)", r.lam1, r.lam2);
        else PASS("shell [[2,1],[1,2]] eigenvalues = (3,1)");
    }

    return fails;
}

/* ─── 6. Fibonacci-spline search ──────────────────────────────── */

static int test_fib_spline(void)
{
    int fails = 0;

    /* Build a small database of 2D unit vectors */
    const int N = 100;
    const int D = 2;
    const int K = 5;

    double* db = (double*)malloc(N * D * sizeof(double));
    srand(777);
    for (int i = 0; i < N; i++) {
        double angle = 2.0 * M_PI * i / N;
        db[i*D+0] = cos(angle);
        db[i*D+1] = sin(angle);
    }

    /* Query = first database vector → should find index 0 as top */
    double query[2] = { db[0], db[1] };
    SearchResult results[K];

    double t0 = now_sec();
    fib_spline_search(query, db, N, D, K, results);
    double dt = now_sec() - t0;

    printf("  fib_spline search: %d vectors, top-%d in %.2f ms\n", N, K, dt*1000);
    printf("    top result: idx=%d sim=%.6f\n", results[0].index, results[0].similarity);

    if (results[0].index != 0)
        FAIL("fib_spline top-1 recall", "expected 0, got %d", results[0].index);
    else PASS("fib_spline top-1 correct index");

    if (fabs(results[0].similarity - 1.0) > 1e-10)
        FAIL("fib_spline top-1 similarity", "%.6f", results[0].similarity);
    else PASS("fib_spline top-1 similarity = 1.0");

    /* Query near element 50 → should find 50 in top-K */
    double q50[2] = { db[50*D] + 0.01, db[50*D+1] + 0.01 };
    fib_spline_search(q50, db, N, D, K, results);

    int found50 = 0;
    for (int i = 0; i < K; i++) {
        if (results[i].index == 50) { found50 = 1; break; }
    }
    if (!found50)
        FAIL("fib_spline recall idx=50", "not in top-%d", K);
    else PASS("fib_spline recall idx=50 in top-K");

    free(db);
    return fails;
}

/* ─── Main ─────────────────────────────────────────────────────── */

int main(void)
{
    printf("═══ Deadband Framework C Library — Test Suite ═══\n\n");

    double t_total = now_sec();
    int total_fails = 0;

    printf("── 1. Eisenstein Snap ──\n");
    total_fails += test_eisenstein();

    printf("\n── 2. HPDF Sampling ──\n");
    total_fails += test_hpdf();

    printf("\n── 3. /360 Arithmetic ──\n");
    total_fails += test_div360();

    printf("\n── 4. Berlekamp-Massey ──\n");
    total_fails += test_bma();

    printf("\n── 5. Shell Decompose ──\n");
    total_fails += test_shell();

    printf("\n── 6. Fibonacci-Spline Search ──\n");
    total_fails += test_fib_spline();

    double elapsed = now_sec() - t_total;
    printf("\n═══ Results: %d failures | %.1f ms total ═══\n", total_fails, elapsed*1000);

    return total_fails;
}
